mixer/adapter/signalfx/metrics.go

```diff
@@ -15,15 +15,72 @@
 package signalfx
 
 import (
-	"errors"
+	"context"
 	"time"
 
+	me "github.com/hashicorp/go-multierror"
+	"github.com/signalfx/golib/errors"
+	"github.com/signalfx/golib/sfxclient"
+
 	"istio.io/istio/mixer/adapter/signalfx/config"
 	"istio.io/istio/mixer/pkg/adapter"
 	"istio.io/istio/mixer/template/metric"
 )
 
-func (h *handler) processMetric(conf *config.Params_MetricConfig, inst *metric.Instance) error {
+type metricshandler struct {
+	env                adapter.Env
+	ctx                context.Context
+	scheduler          *sfxclient.Scheduler
+	sink               *sfxclient.HTTPSink
+	registry           *registry
+	metricTypes        map[string]*metric.Type
+	intervalSeconds    uint32
+	metricConfigByName map[string]*config.Params_MetricConfig
+}
+
+func (mh *metricshandler) InitMetrics() error {
+	mh.scheduler = sfxclient.NewScheduler()
+	mh.scheduler.Sink = mh.sink
+	mh.scheduler.ReportingDelay(time.Duration(mh.intervalSeconds) * time.Second)
+
+	mh.scheduler.ErrorHandler = func(err error) error {
+		return mh.env.Logger().Errorf("Error sending datapoints to SignalFx: %s", err.Error())
+	}
+
+	mh.registry = newRegistry(5 * time.Minute)
+	mh.scheduler.AddCallback(mh.registry)
+
+	mh.env.ScheduleDaemon(func() {
+		err := mh.scheduler.Schedule(mh.ctx)
+		if err != nil {
+			if ec, ok := err.(*errors.ErrorChain); !ok || ec.Tail() != context.Canceled {
+				_ = mh.env.Logger().Errorf("SignalFx scheduler shutdown unexpectedly: %s", err.Error())
```

Comment:
Here and elsewhere, no need to utter "SignalFX" in the log message, the logger will inject the handler/adapter name automatically.